### PR TITLE
API for setting OCSP Callback

### PR DESCRIPTION
**Pull request by persiaAziz**

### CODE REVIEW

**Comment by persiaAziz** [comment]
You need a cert to initialize the ocsp data, but you don't need a cert to create a context. Again the data to be stapled does not get initialized in the API that creates context.  I think it should be used by people who needs ocsp stapling. There is no need to initialize stapled data if that is not required.

**Comment by persiaAziz** [comment]
Another scenario is, a plugin might choose to have it's own ocsp mechanism, there is no point in initializing the stapled data in that case too.

File: proxy/InkAPI.cc [hunk 1]
```diff
@@ -9682,3 +9683,16 @@ TSHttpTxnPostBufferReaderGet(TSHttpTxn txnp)
   HttpSM *sm = (HttpSM *)txnp;
   return (TSIOBufferReader)sm->get_postbuf_clone_reader();
 }
+
+tsapi TSReturnCode
+TSSslInitOCSPStapling(TSSslContext ctx, X509 *cert, char *certname)
```
Comment (bryancall) [inline proxy/InkAPI.cc]: Why do the docs differ from the function added?

**Comment by calavera** [comment]
I built OCSP stapling into Netlify's plugin about a year ago because this API was not available. This is my very personal opinion from an usability point of view:

For me, it makes more sense to have an explicit call to initialize the stapling callback, than hiding the call inside a different api. When I wrote our stapling code, I was actually looking for a hook like this into ATS, and it would have never crossed my mind that it could be inside a call like TSSslServerContextCreate.

After we deployed our implementation, we ran into issues with some CAs not replying fast enough to stapling refresh calls, which cause our certificates hooks to slow down SNI RTT. We added a feature flag to our plugin to be able to enable and disable OCSP stapling on demand. I know this could also be done with @zwoop's approach, but it'd be more hidden. This is how the initialization in our plugin looks like:

```
if (ocsp::stapling_enabled()) {
  SSL_CTX_set_tlsext_status_cb(ssl_ctx, ocsp::ssl_callback_ocsp_stapling);
   if (!ocsp::ssl_stapling_init_cert(ssl_ctx, cert, servername.c_str())) {
     TSDebug("netlify_ocsp", "fail to configure SSL_CTX for OCSP Stapling info for certificate at %s", servername.c_str());
  }
}
```

The migration to this new API would be very straight forward as you can imagine.

I'm not familiar with old API decisions, so no idea why/how the old API is broken, so I cannot comment on that.

**Comment by zwoop** [comment]
[approve ci clang-format]

**Comment by zwoop** [comment]
Thanks for humoring the old curmudgeon. +1!

**Pull request merged**